### PR TITLE
✨ Source S3: Add reading files inside zip archive

### PR DESCRIPTION
**Pull request by tolik0**
## What
In this PR, we've enhanced the capability of our S3 stream reader to efficiently access and read files within ZIP archives stored in S3. The key highlights include:

- **Central Directory Parsing:** Introduced logic to read and parse the Central Directory of ZIP archives, which provides metadata about the files contained within.
- **Direct File Access:** Leveraged the offset information from the Central Directory to directly seek and read individual files from the ZIP archive without the need to decompress the entire archive.

This implementation ensures a faster and more efficient way to access files within ZIP archives in S3, which is especially beneficial for large archives.

**Closes:**
https://github.com/airbytehq/airbyte/issues/17021

## How
The process of reading files from a ZIP archive stored in S3 is divided into two main steps:

**1. Reading the Central Directory of the ZIP Archive:**

- The Central Directory (CD) is a crucial part of the ZIP file format. It contains metadata about all the files and directories stored inside the ZIP archive.
- Our implementation begins by reading the Central Directory to gather information about the contents of the ZIP archive. The ZipFileHandler is responsible for this task.
- By parsing the Central Directory, we obtain essential details about each file inside the ZIP, such as its name, compression method, compressed size, and most importantly, its offset within the ZIP archive. The offset is crucial because it tells us where the file's data starts in the ZIP archive.

**2. Reading Individual Files Using Their Offset:**

- Once we have the offset of a file from the Central Directory, we can directly seek to that position in the ZIP archive to start reading the file's data.
- The RemoteFileInsideArchive class represents each file inside the ZIP archive. It stores metadata about the file, including its offset, which is used to read the file.
- When a request is made to read a specific file from the ZIP archive, our implementation seeks to the file's offset in the ZIP archive and begins reading the file's data from that position. This direct seek-and-read approach ensures efficient and fast access to individual files without the need to decompress or read the entire ZIP archive.

In summary, our implementation efficiently reads files from a ZIP archive in S3 by first parsing the Central Directory to gather metadata and then using the offset information to directly access and read individual files.

### CODE REVIEW

**Review by davydov-d — CHANGES_REQUESTED** [review]
Please add tests, preferably CAT with the new archive type

**Comment by airbyte-oss-build-runner** [comment]
##  source-s3 test report (commit [`e56c48c8b5`](https://github.com/airbytehq/airbyte/pull/31340/commits/e56c48c8b51f8b0e4de2e929f22fe251706e94da)) - ✅

⏲️  Total pipeline duration: 14mn55s 

| Step                                                      | Result   |
|:----------------------------------------------------------|:---------|
| Build source-s3 docker image for platform(s) linux/x86_64 | ✅        |
| Unit tests                                                | ✅        |
| Acceptance tests                                          | ✅        |
| Code format checks                                        | ✅        |
| Validate metadata for source-s3                           | ✅        |
| Connector version semver check                            | ✅        |
| Connector version increment check                         | ✅        |
| QA checks                                                 | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/tolik0_source-s3_add-reading-inside-zip/1697198826/e56c48c8b51f8b0e4de2e929f22fe251706e94da/source-s3/4.1.0/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:e56c48c8b51f8b0e4de2e929f22fe251706e94da)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-s3 test
```

**Review by davydov-d — APPROVED** [review]
:clap:

File: airbyte-integrations/connectors/source-s3/source_s3/v4/zip_reader.py [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+        # Mapping between uncompressed and compressed offsets for quick seeking
+        self.offset_map = {0: self.file_start, self.uncompressed_size: self.file_start + self.compressed_size}
+
+    def _calculate_actual_start(self, file_start):
+        """
+        Calculate the actual start of the file content, skipping the ZIP headers.
+        """
+        self._file.seek(file_start + 26)  # Offset to the length fields in the ZIP header
+        name_len, extra_len = struct.unpack("<HH", self._file.read(4))
+        return file_start + 30 + name_len + extra_len
```
Comment (bazarnov) [inline airbyte-integrations/connectors/source-s3/source_s3/v4/zip_reader.py:140]: Could you please give more info about these hardcoded ints: `26` and `30` ?

**Review by bazarnov — APPROVED** [review]

**Comment by airbyte-oss-build-runner** [comment]
##  source-s3 test report (commit [`7217969808`](https://github.com/airbytehq/airbyte/pull/31340/commits/721796980873373d92356e1101b2255a26bc54db)) - ❌

⏲️  Total pipeline duration: 20mn20s 

| Step                                                      | Result   |
|:----------------------------------------------------------|:---------|
| Build source-s3 docker image for platform(s) linux/x86_64 | ✅       |
| Unit tests                                                | ✅       |
| Acceptance tests                                          | ❌       |
| Code format checks                                        | ✅       |
| Validate metadata for source-s3                           | ✅       |
| Connector version semver check                            | ✅       |
| Connector version increment check                         | ✅       |
| QA checks                                                 | ✅       |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/tolik0_source-s3_add-reading-inside-zip/1697585444/721796980873373d92356e1101b2255a26bc54db/source-s3/4.1.0/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:721796980873373d92356e1101b2255a26bc54db)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-s3 test
```

**Comment by airbyte-oss-build-runner** [comment]
##  source-s3 test report (commit [`5f74715bbe`](https://github.com/airbytehq/airbyte/pull/31340/commits/5f74715bbe569ce9886c9d967c0739e65a91fcd8)) - ✅

⏲️  Total pipeline duration: 21mn00s 

| Step                                                      | Result   |
|:----------------------------------------------------------|:---------|
| Build source-s3 docker image for platform(s) linux/x86_64 | ✅       |
| Unit tests                                                | ✅       |
| Acceptance tests                                          | ✅       |
| Code format checks                                        | ✅       |
| Validate metadata for source-s3                           | ✅       |
| Connector version semver check                            | ✅       |
| Connector version increment check                         | ✅       |
| QA checks                                                 | ✅       |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/tolik0_source-s3_add-reading-inside-zip/1697616590/5f74715bbe569ce9886c9d967c0739e65a91fcd8/source-s3/4.1.0/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:5f74715bbe569ce9886c9d967c0739e65a91fcd8)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-s3 test
```
